src/arena.rs

```diff
@@ -288,6 +288,34 @@ impl<T> Arena<T> {
             slot: 0,
         }
     }
+
+    /// Remove all entries in the `Arena` which don't satisfy the provided predicate.
+    pub fn retain<F: FnMut(Index, &mut T) -> bool>(&mut self, mut f: F) {
+        for (i, entry) in self.storage.iter_mut().enumerate() {
+            if let Entry::Occupied(occupied) = entry {
+                let index = Index {
+                    slot: i as u32,
+                    generation: occupied.generation,
+                };
+
+                if !f(index, &mut occupied.value) {
+                    // We can replace an occupied entry with an empty entry with the
+                    // same generation. On next insertion, this generation will
+                    // increment.
+                    *entry = Entry::Empty(EmptyEntry {
+                        generation: occupied.generation,
+                        next_free: self.first_free,
+                    });
+
+                    // The next time we insert, we can re-use the empty entry we
+                    // just created. If another removal happens before then, that
+                    // entry will be used before this one (FILO).
+                    self.first_free = Some(FreePointer::from_slot(index.slot));
+                    self.len = self.len.checked_sub(1).unwrap_or_else(|| unreachable!());
```

Comment:
We know fairly trivially that this len will never underflow -- we just verified that we have an item in the map. Can you add a comment pointing that out for future adventurers?

Reply:
Sure; I actually copied that directly from `remove`, which is why that comment is there. I'll change the comment to reflect its new purpose.

Reply:
Ahah, maybe we should add a comment to both cases! I think the comment about `first_free` is good, mostly just making a new block comment for the `self.len` assignment.

Reply:
Sure! That should be done as a different PR though I think. Comment updated here, anyhow.